content/docs/render-props.md

```diff
@@ -0,0 +1,287 @@
+
+One interesting thing to note about render props is that you can implement most [higher-order components](/docs/higher-order-components.html) (HOC) using a regular component with a render prop. For example, if you would prefer to have a `withMouse` HOC instead of a `<Mouse>` component, you could easily create one using a regular `<Mouse>` with a render prop:
+
+```js
+// If you really want a HOC for some reason, you can easily
+// create one using a regular component with a render prop!
+function withMouse(Component) {
+  return class extends React.Component {
+    render() {
+      return (
+        <Mouse render={mouse => (
+          <Component {...this.props} mouse={mouse} />
+        )}/>
+      );
+    }
+  }
+}
+```
+
+So using a render prop makes it possible to use either pattern.
+
+## Caveats
+
+### Be careful when using Render Props with React.PureComponent
+
+Using a render prop can negate the advantage that comes from using [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) if you create the function inside a `render` method. This is because the shallow prop comparison will always return `false` for new props, and each `render` in this case will generate a new value for the render prop.
+
+For example, continuing with our `<Mouse>` component from above, if `Mouse` were to extend `React.PureComponent` instead of `React.Component`, our example would look like this:
+
+```js
+class Mouse extends React.PureComponent {
+  // Same implementation as above...
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+
+        {/*
+          This is bad! The value of the `render` prop will
+          be different on each render.
+        */}
+        <Mouse render={mouse => (
+          <Cat mouse={mouse} />
+        )}/>
+      </div>
+    );
+  }
+}
+```
+
+In this example, each time `<MouseTracker>` renders, it generates a new function as the value of the `<Mouse render>` prop, thus negating the effect of `<Mouse>` extending `React.PureComponent` in the first place!
+
+To get around this problem, you can sometimes define the prop as an instance method, like so:
+
+```js
+class MouseTracker extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // This binding ensures that `this.renderTheCat` always refers
+    // to the *same* function when we use it in render.
+    this.renderTheCat = this.renderTheCat.bind(this);
+  }
+
+  renderTheCat(mouse) {
+    return <Cat mouse={mouse} />
+  }
+
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <Mouse render={this.renderTheCat} />
+      </div>
+    );
+  }
+}
+```
+
+In cases where you cannot bind the instance method ahead of time in the constructor (e.g. because you need to close over the component's props and/or state) you should extend `React.Component` instead.
```

Comment:
How about mentioning some other ways of writing render props at the end?

---

## Using Props Other Than `render`

Although the examples above use `render`, any prop can be used to implement the render prop pattern. For example, the component below passes a function as the `children` prop:

```jsx

<WithTitleAsync url="/my/api/books/123">
  { ({ title }) => <h1>{title}</h1> }
</WithTitleAsync>

```

```jsx

class WithTitleAsync extends React.Component {
  constructor(props) {
    super(props);
    this.state = { title: null };
  }

  componentDidMount() {
    fetch(this.props.url)
      .then(
        ({ title }) => { this.setState({ title }) },
        (error) => { this.setState({ error }) }
      )
  }

  render () {
    if (this.state.error) {
      return <span>Title Not Found</span>;
    } else if (!this.state.title) {
      return <span>Getting Title...</span>
    } else {
      return this.props.children(this.state.title);
    }
  }
};

WithTitleAsync.propTypes = {
  children: PropTypes.func.isRequired,
};
```

## Optimizing Render Props

If you've benchmarked your app and discovered that a render prop declaration is a performance bottleneck, here are some techniques to optimize:

- Move inline functions to be declared outside the render.

**Before: Inline Render Prop**
```jsx
class ABC extends React.Component {
  render() {
    <MyComponent render={() => <h1>Hello World</h1>}
  }
}
```

**After: Declared Outside Class**
```jsx
class ABC extends React.Component {
  render() {
    return <MyComponent render={helloWorld} />
  }
}

function helloWorld() {
  return <h1>Hello World</h1>
}
```

**After: Using Getters**
```jsx
class ABC extends React.Component {
  get helloWorld() {
    return <h1>Hello World</h1>
  }

  render() {
    return <MyComponent render={this.helloWorld} />
  }
}
```

- Memoize the render function using a utility like [Recompose](https://github.com/acdlite/recompose) or [Lodash.Memoize](https://lodash.com/docs/4.17.4#memoize)